src/runtime/pmix_init.c

```diff
@@ -328,8 +327,8 @@ int pmix_rte_init(uint32_t type, pmix_info_t info[], size_t ninfo, pmix_ptl_cbfu
     ret = pmix_hotel_init(&pmix_globals.notifications, pmix_globals.max_events, pmix_globals.evbase,
                           pmix_globals.event_eviction_time, _notification_eviction_cbfunc);
     PMIX_CONSTRUCT(&pmix_globals.nspaces, pmix_list_t);
-    PMIX_CONSTRUCT(&pmix_globals.keyindex, pmix_pointer_array_t);
-    pmix_pointer_array_init(&pmix_globals.keyindex, 1024, INT_MAX, 128);
+    PMIX_CONSTRUCT(&pmix_globals.keyindex, pmix_keyindex_t);
+    pmix_pointer_array_init(pmix_globals.keyindex.table, 1024, INT_MAX, 128);
```

Comment:
Just a nit - why not just put the array_init call in the constructor?

Reply:
I agree, it's not great. Since we can't include the initial size during construct (and we have to be careful about now we use shared-memory), I left it as a separate step.

Reply:
Understood - but the initial size truly doesn't matter. You could set it to anything - say 256 - as a lower bound since we know we have a couple of hundred defined attributes anyway. You have no upper bound, and so it will simply realloc when something beyond it's current size gets set - and then realloc again if/when that boundary gets hit.

Even if you leave it at 1024, it is so small compared to the shmem footprint that it is in the noise. So you might as well just put it in the constructor.

Reply:
The problem is that we can't realloc using the gds/shmem's TMA.

Reply:
Ouch - in that case, we have a real design problem. Much of the system relies on the ability to realloc as the job scales. Perhaps we need to start by looking at that problem?

Reply:
We were aware of this limitation. That's why I'm careful about how I size data structures. Supporting realloc would require a massive change to gds/shmem's memory allocator. To support realloc, we would have to implement a proper memory allocator, not just a simple linear slab allocator.

Reply:
We probably should chat about this by voice instead of here - would have a time?

Reply:
Oops. My comment was lost before I resolved this. Let me recreate what was lost:

Offline conversion: this limitation is understood, but we need to have a broader conversion about the current limitations of the gds/shmem memory allocation subsystem and its implications.

Reply:
Addressed in #3214.